synapse/cores/sortedram.py

```diff
@@ -0,0 +1,222 @@
+            if not isint(valu[0]):
+                return 0
+            min = (None, None, valu[0], None)
+
+        max = None
+        if valu[1] is not None:
+            if not isint(valu[1]):
+                return 0
+            max = (None, None, valu[1], None)
+
+        rowsbyprop = self.irowsbyprop.get(prop)
+        if not rowsbyprop:
+            return 0
+
+        i = 0
+        for _ in rowsbyprop.irange(minimum=min, maximum=max, inclusive=(True, min and max)):
+            i += 1
+        return i
+
+    def _rowsByRange(self, prop, valu, limit=None):
+        min = None
+        if valu[0] is not None:
+            if not isint(valu[0]):
+                return []
+            min = (None, None, valu[0], None)
+
+        max = None
+        if valu[1] is not None:
+            if not isint(valu[1]):
+                return []
+            max = (None, None, valu[1], None)
+
+        rows = []
+        rowsbyprop = self.irowsbyprop.get(prop)
+        if not rowsbyprop:
+            return rows
+
+        for i, row in enumerate(rowsbyprop.irange(minimum=min, maximum=max, inclusive=(True, min and max))):
+            if limit is not None and i > limit:
+                return rows
+            rows.append(row)
+        return rows
+
+    def _sizeByGe(self, prop, valu, limit=None):
+        return self._sizeByRange(prop, (valu, None), limit=limit)
+
+    def _rowsByGe(self, prop, valu, limit=None):
+        return self._rowsByRange(prop, (valu, None), limit=limit)
+
+    def _sizeByLe(self, prop, valu, limit=None):
+        return self._sizeByRange(prop, (None, valu), limit=limit)
+
+    def _rowsByLe(self, prop, valu, limit=None):
+        return self._rowsByRange(prop, (None, valu), limit=limit)
+
+    def _addRows(self, rows):
+        for row in rows:
+            self.rowsbyid[row[0]].add(row)
+            if isint(row[2]):
+                rowsbyprop = self.irowsbyprop
+            else:
+                rowsbyprop = self.srowsbyprop
+            rowsbyprop[row[1]].add(row)
+            self.rowsbyvalu[ (row[1],row[2]) ].add(row)
+
+    def _delRowsById(self, ident):
+        for row in self.rowsbyid.pop(ident,()):
+            self._delRawRow(row)
+
+    def _delRowsByIdProp(self, iden, prop):
+        rows = [ row for row in self.rowsbyid.get(iden) if row[1] == prop ]
+        [ self._delRawRow(row) for row in rows ]
+
+    def _getRowsByIdProp(self, iden, prop):
+        return [ row for row in self.rowsbyid.get(iden) if row[1] == prop ]
+
+    def _delRowsByProp(self, prop, valu=None, mintime=None, maxtime=None):
+        for row in self.getRowsByProp(prop,valu=valu,mintime=mintime,maxtime=maxtime):
+            self._delRawRow(row)
+
+    def _delRawRow(self, row):
+
+        byid = self.rowsbyid.get(row[0])
+        if byid != None:
+            byid.discard(row)
+
+        if isint(row[2]):
+            rowsbyprop = self.irowsbyprop
+        else:
+            rowsbyprop = self.srowsbyprop
```

Comment:
this type switching is a little tedious, but required because `sortedcontainers.SortedSet` cannot mix key types. so, we keep strings separate from ints.